### PR TITLE
Avoid race conditions when building linker tests

### LINKED ISSUE #40398: Trimming tests fail sometimes with race condition
```
/Users/runner/work/1/s/.dotnet/sdk/5.0.100-preview.8.20362.3/Microsoft.Common.CurrentVersion.targets(4187,5): error MSB3026: Could not copy "/Users/runner/work/1/s/artifacts/obj/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll" to "/Users/runner/work/1/s/artifacts/bin/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/obj/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyInjection/src/Microsoft.Extensions.DependencyInjection.csproj] [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/Microsoft.Extensions.Options.TrimmingTests.proj]
/Users/runner/work/1/s/.dotnet/sdk/5.0.100-preview.8.20362.3/Microsoft.Common.CurrentVersion.targets(4187,5): error MSB3026: Could not copy "/Users/runner/work/1/s/artifacts/obj/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll" to "/Users/runner/work/1/s/artifacts/bin/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/obj/Microsoft.Extensions.DependencyInjection/net5.0-Release/Microsoft.Extensions.DependencyInjection.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyInjection/src/Microsoft.Extensions.DependencyInjection.csproj] [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/Microsoft.Extensions.Options.TrimmingTests.proj]
/Users/runner/work/1/s/eng/testing/linker/trimmingTests.targets(102,5): error MSB3073: The command ""/Users/runner/work/1/s/.dotnet/dotnet" build /t:PublishTrimmed /nr:false /warnaserror -p:configuration=Release " exited with code -1. [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.Options/tests/TrimmingTests/Microsoft.Extensions.Options.TrimmingTests.proj]
```

It seems like 2 tests have a P2P to Microsoft.Extensions.DependencyInjection and for some reason it is trying to copy it from obj to bin at the same time.

Here is the binlog of the failure:
https://dev.azure.com/dnceng/_apis/resources/Containers/4760213/Logs_Build_OSX__x64_release_Runtime_Release?itemPath=Logs_Build_OSX__x64_release_Runtime_Release%2FRelease%2FTrimmingTests.binlog

Failed here: https://dev.azure.com/dnceng/public/_build/results?buildId=758123&view=logs&j=94422687-3fa3-516b-cf3b-23eb810befab&t=0015de85-1e1c-5c80-f8e6-95054477de01

cc: @eerhardt @joperezr

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/runtime/issues/40398
Depends on https://github.com/dotnet/arcade/pull/6936

Changes:
- Use static graph restore for linker tests
- Use in-proc MSBuild task to publish linker test projects
- Enabled parallel building of linker tests
- Code clean up...

Filed https://github.com/dotnet/runtime/issues/48027 to fix an unnecessary excessive amount of NuGet invocation.

### CODE REVIEW

File: eng/testing/linker/trimmingTests.targets
Comment (joperezr) [inline eng/testing/linker/trimmingTests.targets:97]: What is the MSBuildRestoreSessionId for?

File: eng/testing/linker/trimmingTests.targets
Comment (joperezr) [inline eng/testing/linker/trimmingTests.targets:100]: I assume so, but by movint this in-proc do we still get to pick the version of the linker that is used to trim the apps? What happens if the parent msbuild already has the linker task loaded into it (since it links out the shared framework) and the versions of the linker we use for runtime and the framework are different?

File: eng/testing/linker/trimmingTests.targets
Comment (safern) [reply eng/testing/linker/trimmingTests.targets:97]: IIRC, this is for the static graph restore

Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:97]: That doesn't have anything to do with static graph restore. It's a property that is set when restore is invoked via MSBuild so that conditions can be applied on that property to determine if the current node is used for restore:

https://github.com/dotnet/msbuild/blob/01ceb4e6dadb3fa41c90626899d68d640aa22db4/src/MSBuild/XMake.cs#L1420-L1423

File: eng/testing/linker/trimmingTests.targets
Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:100]: Good point. To guarantee that, the `ILLinkTasksAssembly` property which determines the assembly to use for the ILLink task, would either need to be passed in globally or as early as possible (i.e. either in Arcade's Build.proj or in dotnet/runtime's Build.proj). @dsplaisted is there a better way to control which ILLink version the sdk uses? I noticed that the sdk imports the ILLink sdk which makes me wonder if it is retrieved from the sdk directory or via nuget via an implicit global.json sdk value? https://github.com/dotnet/sdk/blob/4191572924db60039c2b746f7d261070a056932e/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets#L14

File: eng/testing/linker/trimmingTests.targets
Comment (eerhardt) [inline eng/testing/linker/trimmingTests.targets:97]: Do we still need `@(CliEnvironment)` now that we are invoking MSBuild tasks?

File: src/libraries/tests.proj
Comment (eerhardt) [inline src/libraries/tests.proj:267]: When I originally added this, I also added a `ContinueOnError` property. Can that be removed now?

https://github.com/dotnet/runtime/commit/254ef0f7f7f429ec238735fe6132805e3c38a19f#diff-5ef2fd16f58de91bff2f05ad601ab6ff000c0febabb2312d232f20cfa132a71bR14

Comment (joperezr) [reply src/libraries/tests.proj:267]: Looks like test projects already have a ErrorAndContinue so yeah we might want to remove the ContinueOnError specific for Trimming and just behave the same as other regular test projects.

File: eng/testing/linker/trimmingTests.targets
Comment (dsplaisted) [reply eng/testing/linker/trimmingTests.targets:100]: The ILLink MSBuild SDK (Microsoft.NET.ILLink.Tasks) is part of the .NET SDK, it's only a separate MSBuild SDK for codeflow / engineering reasons.  So it should use the ILLink that shipped with the .NET SDK unless you override it.

Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:96]: I assumed that such `<Import Project="x.props" Sdk="x" />` use the NuGetSdkResolver. Does a fallback exist that looks into the sdk directory first before probing the nuget cache? cc @nkolev92

Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:96]: I submitted https://github.com/dotnet/arcade/pull/6936 to arcade to make sure that the right linker is used (which is very similar to the compiler package opt-in).

Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:96]: @joperezr @eerhardt the fix to guarantee that a newer ILLink package is used over the one that ships with the SDK is in dotnet/arcade#6936 but that one requires a newer SDK build because ILLink props and targets aren't sequenced correctly, which will be fixed by https://github.com/dotnet/sdk/pull/15801. I don't know yet if we can still get that change into preview 1.

Comment (dsplaisted) [reply eng/testing/linker/trimmingTests.targets:96]: This generally won't use the NuGet Sdk Resolver: `<Import Project="x.props" Sdk="x" />`

For the NuGet resolver to be used, the version for the SDK to use has to be specified.  It might be possible to do that in a global.json folder, but I think the .NET SDK resolver would still take precedence.

Comment (ViktorHofer) [reply eng/testing/linker/trimmingTests.targets:96]: Got it. I also found https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019#how-project-sdks-are-resolved which goes into a bit more detail.

I just tried specify a version for ILLink in an app's global.json but it isn't honored presumably because the .NET SDK resolver takes precedence (as you pointed out above).

**Comment by joperezr** [comment]
@ViktorHofer are we waiting for the linker version PR to get merged on Arcade first in order to merge this? If this is not currently blocked, it would be good to get it in since I'm currently working on adding a browser-wasm leg for the trimming tests, and it will be much easier to do it once we publish the test apps in-proc in order to be able to import all of the current infra for building the wasm bundle.

**Comment by ViktorHofer** [comment]
To make sure that the right linker is always used we require a newer SDK as ILLink imports were wrongly sequenced. That said, the right linker is currently used and I don't think we should block the PR on that. Just need an approval and then it's ready to be merged.

**Review by eerhardt — APPROVED** [review]
This looks reasonable to me. Thanks for helping out here, @ViktorHofer!

**Comment by ViktorHofer** [comment]
The build hang for 2h during downloading CoreCLR artifacts. Pinged @MattGal about this offline.

**Pull request merged**
